OpenSim/Simulation/OpenSense/ModelCalibrator.h

```diff
@@ -0,0 +1,89 @@
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+#include <OpenSim/Common/Object.h>
+#include <OpenSim/Simulation/osimSimulationDLL.h>
+#include <Simbody.h>
+namespace OpenSim {
+
+class Model;
+//=============================================================================
+//=============================================================================
+/**
+ * A class implementing a set of parameters describing how to calibrate
+ * on a model to match placement of Sensors (IMUs)
+ *
+ *
+ * @author Ayman Habib
+ */
+class OSIMSIMULATION_API ModelCalibrator : public Object {
+    OpenSim_DECLARE_CONCRETE_OBJECT(ModelCalibrator, Object);
+
+public:
+    //==============================================================================
+    // PROPERTIES
+    //==============================================================================
+    OpenSim_DECLARE_PROPERTY(model_file_name, std::string,
+            "Name/path to the xml .osim file used to load a model to be "
+            "calibrated.");
+
+    OpenSim_DECLARE_PROPERTY(base_imu_label, std::string,
+            "The label of the base IMU in the orientations_file used to account "
+            "for the heading difference between the sensor data and the forward "
+            "direction of the model. Leave blank if no heading correction is applied.");
+
+    OpenSim_DECLARE_PROPERTY(base_heading_axis, std::string,
+            "The axis of the base IMU that corresponds to its heading "
+            "direction. Options are 'x', '-x', 'y', '-y', 'z' or '-z'.");
+
+    OpenSim_DECLARE_PROPERTY(sensor_to_opensim_rotations, SimTK::Vec3,
+            "Space fixed rotations from IMU Space (typically z up) to OpenSim "
+            "(Y up, X forward)");
+
+    OpenSim_DECLARE_PROPERTY(calibration_file_name, std::string,
+            "Name/path to a .sto file of sensor frame orientations as "
+            "quaternions to be used for calibration.");
+
+public:
+    virtual ~ModelCalibrator();
+    ModelCalibrator();
+    ModelCalibrator(const std::string& setupFile);
+    bool run(bool visualizeResults = false);
+    void setModel(Model& aModel) { _model = &aModel; };
+    Model& getModel() const { return *_model; }
```

Comment:
I had doubts on if this returned the original model or the calibrated version at the end. Not sure if we could do something here to ensure the User that the returned model is the calibrated version. Also, I called .getModel() before run(), to test if it returned anything, and I got segFault that crashed Matlab (WHOOPS ;). 

Perhaps we could change the method name to .getCalibratedModel() to convey that this is indeed the calibrated model. And, if called before run(), an exception occurs. Just a thought.

Reply:
Done